### ISSUE TITLE
Deprecate iPython 2 and v3 notebook format

### ISSUE THREAD

**Issue by jakirkham** [paulgb/runipy#78]
I think it is time to deprecate iPython 2 support. Here are few reasons.

1. It is not going to see any more releases (including bug fixes) see comment here ( https://github.com/ipython/ipython/issues/8657#issuecomment-139008775 ).
2. Its last release is nearly two years old ( https://github.com/ipython/ipython/releases/tag/rel-2.4.1 ).
3. It uses the old v3 notebook format, which is beginning to cause us problems ( https://github.com/paulgb/runipy/pull/53 ).
4. There will already need to be some work to properly handle installation changes caused by "The Big Split". ( http://blog.jupyter.org/2015/04/15/the-big-split/ ) ( https://github.com/paulgb/runipy/issues/77 )
5. A large number of the issues reported here are caused by recent changes in iPython that have changed its structure or added new interesting features.

In short, the amount of work to handle a less frequently used version iPython is becoming disproportionate relative to its use. I will try to get as many relevant bug fixes in before making a final iPython 2.x release and then it will not be supported further.

I welcome comments to this proposal.

**Comment by aaren** [comment]
If you use `nbformat.read(input_file, as_version=4)` then you can read in older notebook versions as version 4, with nbformat doing the conversion for you.

This is what I do in [notedown](https://github.com/aaren/notedown) - see usage [here][nbformat-ex]. 

[nbformat-ex]: https://github.com/aaren/notedown/blob/master/notedown/main.py#L253-L275

**Comment by jakirkham** [comment]
Thanks for that idea @aaren. I'll take a look into it.

**Comment by jakirkham** [comment]
The more I think about it. The more I am ok with going forward with this deprecation. More frequently I am bombarded with a long list of features that users would like to have that are just very painful if not impossible to support with IPython 2.x. I think it is time for us to acknowledge this break needs to happen.

**Comment by Juanlu001** [comment]
When I first thought about this IPython 2.x felt quite old, but for my surprise 2.4 was relased in January last year. Maybe the IPython developers have some usage statistics or can shed some light on this topic. @Carreau, do you think it's safe to drop support for IPython 2.x nowadays?

**Comment by Carreau** [comment]
> 2.4 was released in January last year

2.4 yes, but 2.0 was released in April 2014, so more than 2 years ago.
2.x late series was just an **exceptional** backport of nbformat so that 2.4 can read notebooks v4 and allow people to read newer notebook it was likely release a few hours before 3.0.
Now this is irrelevant at nbformat is its own package and can be upgraded  independently. 

Unfortunately we don't have that much statistics, but I think that asking people to at least be on 4.x for new features is reasonable, it's soon going to be a year old, and we're about to release 5.0 (For IPython, not yet for the notebook). 

Notebook v4 is definitively the most widespread at least on GitHub. So IMHO requiring IPython 3 or above is reasonable. Personally I would leapfrog 3 and say you require 4+.

Sidenote, we are going to drop Python 2 support in IPython 6.x (not the notebook just for IPython kernel), you will still be able to run IPython 2 kernel with the notebook but you will be stuck with IPython < 6.0.  cf [here](http://python3statement.github.io/)

**Comment by jakirkham** [comment]
Thoughts on this @zonca?

**Comment by zonca** [comment]
I think it is reasonable to drop IPython 2.x support

**Comment by Carreau** [comment]
Note, I think PyPI have download data available on BigQuery so we should be able to get an idea of how many download / month thereis of each version of IPython.

**Comment by Carreau** [comment]
Which I just did:

```
SELECT REGEXP_EXTRACT(file.version, r'^(\d+).\d+') as major_version,  COUNT(file.version) as downloads
FROM (TABLE_DATE_RANGE([the-psf:pypi.downloads], 
                TIMESTAMP('2016-06-26'), 
                TIMESTAMP('2016-08-31'))) 
WHERE file.project == 'ipython'
GROUP BY major_version
ORDER BY major_version
```

https://bigquery.cloud.google.com/results/iron-wave-88322:bquijob_6dc56858_15749835aec

Here is the result:
```
Row	major_version	downloads	 
1	0     			25760	 
2	1     			39188	 
3	2     			63867	 
4	3     			77119	 
5	4     			354899	 
6	5     			836842	 
````

**Comment by jakirkham** [comment]
Wow! I'm a little surprised to see that much usage of even old versions like `0.x` and `1.x.`. Also a bit surprised to see that `2.x` and `3.x` are comparable.

That all being said, I think the numbers speak for themselves there is around an order of magnitude more people on iPython 5.x and 4.x is still alive and well. For a small project like this, I think we are well justified in doing exactly as you have suggested @Carreau and dropping `3.x` and below. Though dropping `2.x` will vastly improve code quality.

**Comment by Carreau** [comment]
My guess is IPython 2.x is for Python 2.6 We can refine the query to know. 

You can make the same query for runipy to know how many users are using Python 2.6.

**Comment by Carreau** [comment]
Notes; I'm working on https://github.com/Carreau/warn to make deprecation easier. 
Still hackish though. Feedback would be welcome.

**Comment by jakirkham** [comment]
Ah, I see.

We don't actually support Python 2.6. Though it may work there. We just don't test against it.
